crates/ide/src/ide/assists/add_to_lambda.rs

```diff
@@ -0,0 +1,84 @@
+//! add an undefined name to the top-level lambda
+use super::{AssistKind, AssistsCtx};
+use crate::def::Name;
+use crate::TextEdit;
+use syntax::ast::{self, AstNode};
+use syntax::{TextRange, TextSize};
+
+pub(super) fn add_to_lambda(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let name = ctx.covering_node::<ast::Ref>()?.token()?;
+    let name = name.text();
+
+    // the whole file should be a lambda
+    let pat = ast::Lambda::cast(ctx.ast.syntax().first_child()?)?
+        .param()?
+        .pat()?;
+
+    // name should be undefined
+    if ctx
+        .db
+        .module(ctx.frange.file_id)
+        .names()
+        .any(|(_, Name { text, .. })| text == name)
+    {
+        return None;
+    };
+
+    let (pos, insert) = if let Some(field) = pat.fields().last() {
+        let pos = field.syntax().text_range().end();
+        let left = pos - TextSize::from(1);
+        // insert before the space if the field ends with a space
+        (
+            if ctx.ast.syntax().text().char_at(left) == Some(' ') {
+                left
+            } else {
+                pos
+            },
+            format!(", {name}"),
+        )
+    } else if let Some(ellipsis) = pat.ellipsis_token() {
+        (ellipsis.text_range().start(), format!("{name}, "))
+    } else {
+        (
+            pat.syntax().text_range().end() - TextSize::from(1),
```

Comment:
Why an unconditional offset?

Reply:
because the last character would be `}`

Reply:
It's not obvious and may break when `}` is not the last token, in case of trivia tokens or incomplete syntax.
I added `Pat::r_curly_token` so you could retrieve its position now.

Reply:
added that

I wasn't able to create one without an `r_curly_token` so I just threw something in the end, the formatting might be off for that branch